Make MyRocks transactions never expire additional rows with TTL while running

laurynas-biveinis opened the pull request:
The TTL row filtering is governed by Rdb_transaction::m_snapshot_timestamp
field. This field sometimes get reset to zero when a snapshot is released,
indicating that it should be reinitialized at the next snapshot, or to the
current timestamp (when a single statement in a multi-statement transaction gets
rolled back). The timestamp going forward means that additional rows may be
hidden in the same transaction if they happened to expire between the two
snapshots. Change it so that additional rows are never expired:

- Only initialize the field once per transaction, when the first snapshot is
  created.
- Do not touch it for single-statement rollbacks in multi-statement
  transactions.
- Rename the field to m_earliest_snapshot_ts to better reflect what it does.
- Change the global transaction list to be sorted by the oldest snapshot
  timestamp, so that the oldest global one can be found by reading the
  transaction at the list head. The transactions without snapshot are inserted
  to the its tail, and reinserted as needed upon acquiring the first snapshot.
  To make reinsertions fast, a third list pointer is maintained that points to
  the newest transaction with a snapshot.
- Extend the SHOW ENGINE ROCKSDB TRANSACTION STATUS output with transactions
  that currently do not have a snapshot, for both cases of having a snapshot
  before and never having had one.
- Add a test to rocksdb.ttl_primary_read_filtering for the behavior
  difference introduced by this commit.

At the same time:
- Initialize m_earliest_snapshot_ts from the binlog HLC instead of the current
  system time.

hermanlee commented:
I think there is a potential issue in rdb_compact_filter.h with the  `calculate_expiration_timestamp()` function. This function relies on RocksDB to return back the oldest snapshot time to be used for compaction. Even though each transaction keeps it own snapshot time, the compaction code would also need to figure out the oldest snapshot time across all transactions.

MyRocks currently relies on RocksDB to maintain this heap.

I wonder if this diff should be based on the RR/RC change to allow the read-only snapshot in RR to be created from the first read-only select. In that case, the snapshot never goes away, and the snapshot time never changes.

For RC, if we want the same behavior, we could also keep the read-only snapshot around just to maintain the timestamp for compaction, but each new read-only query would still acquire its own snapshot.

laurynas-biveinis commented:
@hermanlee If I understand your comment correctly, currently there is a race condition between the compaction filter and the oldest transaction temporarily releasing its snapshot, thus maintaining only a per-transaction timestamp for the per-transaction filtering is not enough, and a transaction snapshot must be kept open constantly once it has been acquired.

laurynas-biveinis commented:
@hermanlee , it seems that each transaction has to have three snapshot fields, even if it does not use them all in all cases:
- the snapshot for the currently-executed statement;
- the RR RO snapshot;
- the snapshot that prevents expiration.

The 3rd one and the 1st one cannot be merged obviously.
Likewise the 2nd one and the 1st one.
The 3rd one and the 2nd one cannot be merged because an RR transaction may execute writes before its first read.

It could be possible to reduce complexity of transaction object maintaining three snapshots by moving the oldest timestamp tracking from RocksDB to MyRocks (have a heap/priority queue of timestamps in Rdb_compact_filter, insert/remove from it on transaction 1st snapshot create/commit&rollback), but a shared data structure that duplicates RocksDB internals might not be the best trade-off here.

laurynas-biveinis commented:
As discussed offline, will use the global transaction list instead. Step 1 of this approach is at https://github.com/facebook/mysql-5.6/pull/1471

sunshine-Chun commented on storage/rocksdb/rdb_compact_filter.h:
```diff
@@ -50,15 +110,13 @@ class Rdb_compact_filter : public rocksdb::CompactionFilter {
     }
 
-    m_expiration_timestamp = oldest_snapshot_timestamp;
+    m_expiration_timestamp = oldest_transaction_timestamp;
```
Hey Laurynas, we recently notice a new issue that will bring inconsistency to primary and secondary in long running transactions. The oldest_transaction_timestamp here is retrieved from snapshot->GetUnixTime(). This timestamp will be different between primary and secondary. Secondary will have a slightly larger timestamp, hence compact more data away than primary. Can we change the oldest_transaction_timestamp from snapshot->GetUnixTime() to mysql_bin_log.get_max_write_hlc_nsec()? This can ensure the consistency between primary and secondary because mysql_bin_log.get_max_write_hlc_nsec() returns the same value if primary and secondary has the same GTID set when taking snapshot.
cc @abhinav04sharma

sunshine-Chun replied:
#1471 has been closed. Can we have a rebase to include that change? Thanks!

laurynas-biveinis replied:
@sunshine-Chun, OK, I will make it use `mysql_bin_log.get_max_write_hlc_nsec()`. I am rebasing (and rewriting) this PR right now.

laurynas-biveinis replied:
@sunshine-Chun , I am not too familiar with the binlog HLC patch. What value should we use if the HLC value is zero? Should we use `snapshot->GetUnixTime()` then?

laurynas-biveinis replied:
@sunshine-Chun, the relationship between binlog and RocksDB timestamps is not clear to me. We initialize the earliest timestamp from binlog, but then the current snapshot timestamps come from the RocksDB snapshots directly. I'd expect the invariant "the current timestamp can be earlier than the earliest transaction timestamp" to hold, but, i.e. `rocksdb_rpl.rocksdb_binlog_ttl` test shows (`result` is the current snapshot timestamp)
```
(lldb) p/x m_earliest_snapshot_ts
(int64_t) 0x00047fc94dc5cd7c
(lldb) p/x result
(const int64_t) 0x0000000066c7143b
```

sunshine-Chun replied:
Here is what these two timestamp represents:
1. Each write transaction has a HLC timestamp assigned to it. The `mysql_bin_log.get_max_write_hlc_nsec()` tells us the lastest HLC that is written to this instance. 
2. The RocksDB timestamp represents the unix time when creating the snapshot. 

In most cases, these two timestamp should be very close in value. In the long running transaction case, when creating snapshot,  `mysql_bin_log.get_max_write_hlc_nsec()` should always be smaller than the RocksDB timestamp, cause no new writes will update the max_write_hlc_nsec at that time.

The reason why we use  `mysql_bin_log.get_max_write_hlc_nsec()` here is that it can help the primary and secondary be consistent at the this value. When we create snapshot in long running transactions in primary and secondary, we can make sure they are on the same GTID set. This also means they will have the same `mysql_bin_log.get_max_write_hlc_nsec()` value. Noted primary and secondary maintains their own `mysql_bin_log.get_max_write_hlc_nsec()` value. But since they have the same committed transactions, they share the same `mysql_bin_log.get_max_write_hlc_nsec()` value. Rocksdb timestamp can not do this cause it will get the unix now in primary and secondary separately. Secondary takes snapshot later than primary.

sunshine-Chun replied:
`mysql_bin_log.get_max_write_hlc_nsec()` is unlikely to be zero. But yeah, `snapshot->GetUnixTime()` seems the only thing we can do to guard the `new_earliest_ts` from zero.

sunshine-Chun replied:
MTR test seems to be failing. Have you rebased it to the lastest branch?

laurynas-biveinis replied:
@sunshine-Chun, thank you, this makes sense. It seems that the "L" in "HLC" is not stored explicitly, thus the binlog HLC is directly comparable to RocksDB timestamps, different resolutions aside. In which case I still cannot understand what is going on with `rocksdb_rpl.rocksdb_binlog_ttl`. The testcase does not inject any artificial timestamps, and yet:
```
$ p result # The current RDB snapshot timestamp
(const int64_t) 1724752351 # OK, resolves to $NOWish timestamp at the second resolution
(lldb) p m_earliest_snapshot_ts # Binlog HLC
(int64_t) 1695877370630958 # Does not resolve to anything reasonable at any resolution
```

sunshine-Chun replied:
I run the test in my local server and got the following number. The `m_earliest_snapshot_ts` looks correct to me. Note, this is the value that hit the assertion `assert(m_earliest_snapshot_ts <= result || result == 0);`
```
(gdb) p m_earliest_snapshot_ts
$1 = 1724827776952667189
(gdb) p result
$2 = 1724827777
```
Let me post the logic to get HLC here.
```
uint64_t HybridLogicalClock::get_next() {
  uint64_t current_hlc, next_hlc = 0;
  bool done = false;

  while (!done) {
    // Get the current wall clock in nanosecond precision
    uint64_t current_wall_clock =
        std::chrono::duration_cast<std::chrono::nanoseconds>(
            std::chrono::high_resolution_clock::now().time_since_epoch())
            .count();

    // get the 'current' internal HLC
    current_hlc = current_.load();

    // Next HLC timestamp is max of current-hlc and current-wall-clock
    next_hlc = std::max(current_hlc + 1, current_wall_clock);

    // Conditionally update the internal hlc
    done = current_.compare_exchange_strong(current_hlc, next_hlc);
  }

  return next_hlc;
}
```

Can try to print the `current_wall_clock` to see if the value is abnormal? This determines the baseline value of the HLC. My guess is that the abnormal value starts from here.

laurynas-biveinis replied:
`current_wall_clock` value is indeed correct, thus the only issue was comparing nanoseconds to seconds directly. Adding the required conversion fixed the tests.

laurynas-biveinis commented:
Non-functional changes in this PR split off to https://github.com/facebook/mysql-5.6/pull/1487

sunshine-Chun commented:
> Non-functional changes in this PR split off to #1487

The change has been committed to the codebase.

laurynas-biveinis committed:
Make MyRocks transactions never expire additional rows with TTL after starting

The TTL row filtering is governed by Rdb_transaction::m_snapshot_timestamp
field. This field sometimes get reset to zero when a snapshot is released,
indicating that it should be reinitialized at the next snapshot, or to the
current timestamp (when a single statement in a multi-statement transaction gets
rolled back). The timestamp going forward means that additional rows may be
hidden in the same transaction if they happened to expire between the two
snapshots. Change it so that additional rows are never expired:

- Only initialize the field once per transaction, when the first snapshot is
  created.
- Do not touch it for single-statement rollbacks in multi-statement
  transactions.
- Rename the field to m_earliest_snapshot_ts to better reflect what it does.
- Change the global transaction list to be sorted by the oldest snapshot
  timestamp, so that the oldest global one can be found by reading the
  transaction at the list head. The transactions without snapshot are inserted
  to the its tail, and reinserted as needed upon acquiring the first snapshot.
  To make reinsertions fast, a third list pointer is maintained that points to
  the newest transaction with a snapshot.
- Extend the SHOW ENGINE ROCKSDB TRANSACTION STATUS output with transactions
  that currently do not have a snapshot, for both cases of having a snapshot
  before and never having had one.
- Add a test to rocksdb.ttl_primary_read_filtering for the behavior
  difference introduced by this commit.

At the same time:
- Initialize m_earliest_snapshot_ts from the binlog HLC instead of the current
  system time.

Note for downstreams: if you don't carry the binlog HLC patch,
binlog_hlc_or_snapshot_ts method simplifies to a snapshot->GetUnixTime call.

laurynas-biveinis commented:
@sunshine-Chun , I have pushed my current version, which is as finished as possible, except for the binlog/RDB timestamp question above. Keeping in draft status until that is resolved.

laurynas-biveinis commented:
> MTR test seems to be failing. Have you rebased it to the lastest 
> branch?

Which tests? Two debug MTR tests are failing with assert RDB 
timestamp > binlog timestamp, as I wrote before, and are the 
reason this PR is still draft.

sunshine-Chun commented:
> MTR test seems to be failing. Have you rebased it to the lastest branch?
> Which tests? Two debug MTR tests are failing with assert RDB timestamp > binlog timestamp, as I wrote before, and are the reason this PR is still draft.

Other than the assertions:
1. `drop_cf_before_show_deadlock_info` needs to be re-recorded.
2. `performance_schema` length mismatch: `+Performance_schema_digest_lost	2374
`
3. `rocksdb_concurrent_delete` result content mismatch
4. sanitize_report, will sent the log via email.

sunshine-Chun commented on storage/rocksdb/rdb_compact_filter.h:
```diff
@@ -50,15 +50,12 @@ class Rdb_compact_filter : public rocksdb::CompactionFilter {
-    m_expiration_timestamp = oldest_snapshot_timestamp;
+    m_expiration_timestamp = oldest_transaction_ts;
 
     if (rdb_is_binlog_ttl_enabled()) {
```
I wonder if we can directly use the rocksdb_binlog_ttl_compaction_timestamp here instead of getting the min between two values.

sunshine-Chun replied:
> I wonder if we can directly use the rocksdb_binlog_ttl_compaction_timestamp here instead of getting the min between two values.

We can't. Because if we have a long running transaction in secondary, the oldest_snapshot_timestamp can be older than the rocksdb_binlog_ttl_compaction_timestamp. We nee the oldest_snapshot_timestamp to prevent compacting away records needed in this long running transcation.

sunshine-Chun commented:
Just think aloud.  Feel free to correct me if I miss anything here.

`m_binlog_ttl_read_filtering_ts` will only be only assigned once per transaction, so it basically behaves like the earliest_snapshot_ts in this change.  If we refresh the snapshot during the long running transaction, the `m_binlog_ttl_read_filtering_ts` won't change its value. The `rocksdb_binlog_ttl_compaction_timestamp` is based on `m_binlog_ttl_read_filtering_ts`, so it won't change too. So it seems with binlog_ttl enabled, we already won't filter out more row when refreshing snapshot.

laurynas-biveinis commented:
> 2. `performance_schema` length mismatch: `+Performance_schema_digest_lost	2374 `
> 3. `rocksdb_concurrent_delete` result content mismatch

Cannot reproduce these two. For the latter, I intermittently get a pre-existing failure:
```
[ 71%] rocksdb.rocksdb_concurrent_delete 'innodb_ddse'  [ fail ]
        Test ended at 2024-08-27 12:24:08

CURRENT_TEST: rocksdb.rocksdb_concurrent_delete
--- /home/laurynas/vilniusdb/fb-mysql/mysql-test/suite/rocksdb/r/rocksdb_concurrent_delete.result	2024-03-19 14:53:38.570572735 +0300
+++ /home/laurynas/vilniusdb/fb-mysql/_build-llvm-debug/mysql-test/var/log/rocksdb_concurrent_delete.reject	2024-08-27 15:24:08.695347028 +0300
@@ -453,9 +453,9 @@
 set debug_sync='now SIGNAL go';
 select * from t1 where id1=1;
 id1	id2	value
+1	2	200
 1	4	1
 1	5	1
-1	2	200
 --End row delete with PRIMARY
 SET SESSION TRANSACTION ISOLATION LEVEL REPEATABLE READ;
 set debug_sync='rocksdb.get_row_by_rowid SIGNAL parked WAIT_FOR go';

mysqltest: Result content mismatch
```

sunshine-Chun commented:
> Just think aloud. Feel free to correct me if I miss anything here.
> 
> `m_binlog_ttl_read_filtering_ts` will only be only assigned once per transaction, so it basically behaves like the earliest_snapshot_ts in this change. If we refresh the snapshot during the long running transaction, the `m_binlog_ttl_read_filtering_ts` won't change its value. The `rocksdb_binlog_ttl_compaction_timestamp` is based on `m_binlog_ttl_read_filtering_ts`, so it won't change too. So it seems with binlog_ttl enabled, we already won't filter out more row when refreshing snapshot.

The current change and binlog ttl feature covers different scenarios. Both are needed. 
`m_expiration_timestamp = min(oldest_transaction_ts, rocksdb_binlog_ttl_compaction_timestamp)`

- When m_expiration_timestamp = oldest_transaction_ts, the current change works. This happens when we have a long running transaction in secondary. 
- When m_expiration_timestamp = rocksdb_binlog_ttl_compaction_timestamp, the binlog ttl feature works. This happens when we have a long running transaction in both primary and secondary.

sunshine-Chun commented:
> > 2. `performance_schema` length mismatch: `+Performance_schema_digest_lost	2374 `
> > 3. `rocksdb_concurrent_delete` result content mismatch
> 
> Cannot reproduce these two. For the latter, I intermittently get a pre-existing failure:
> 
> ```
> [ 71%] rocksdb.rocksdb_concurrent_delete 'innodb_ddse'  [ fail ]
>         Test ended at 2024-08-27 12:24:08
> 
> CURRENT_TEST: rocksdb.rocksdb_concurrent_delete
> --- /home/laurynas/vilniusdb/fb-mysql/mysql-test/suite/rocksdb/r/rocksdb_concurrent_delete.result	2024-03-19 14:53:38.570572735 +0300
> +++ /home/laurynas/vilniusdb/fb-mysql/_build-llvm-debug/mysql-test/var/log/rocksdb_concurrent_delete.reject	2024-08-27 15:24:08.695347028 +0300
> @@ -453,9 +453,9 @@
>  set debug_sync='now SIGNAL go';
>  select * from t1 where id1=1;
>  id1	id2	value
> +1	2	200
>  1	4	1
>  1	5	1
> -1	2	200
>  --End row delete with PRIMARY
>  SET SESSION TRANSACTION ISOLATION LEVEL REPEATABLE READ;
>  set debug_sync='rocksdb.get_row_by_rowid SIGNAL parked WAIT_FOR go';
> 
> mysqltest: Result content mismatch
> ```

Then they are not related to this change.

laurynas-biveinis commented:
> 4. sanitize_report, will sent the log via email.

This was caused by concurrent accesses to `m_read_opts[TABLE_TYPE::USER_TABLE].snapshot`, which appears to be a pre-existing issue and it seems that the current code may crash on `SHOW ENGINE ROCKSDB STATUS` concurrent to i.e. statement rollbacks. I am fixing it as follows, but overall the accesses do not appear to be synchronized properly and I'm making a TODO note to review it:
```
diff --git a/storage/rocksdb/ha_rocksdb.cc b/storage/rocksdb/ha_rocksdb.cc
index 3b5cf2da90c..9ba4860b03a 100644
--- a/storage/rocksdb/ha_rocksdb.cc
+++ b/storage/rocksdb/ha_rocksdb.cc
@@ -5503,6 +5503,10 @@ class Rdb_transaction_impl : public Rdb_transaction {
     if (m_rocksdb_tx[TABLE_TYPE::USER_TABLE]) {
       const rocksdb::Snapshot *const org_snapshot =
           m_rocksdb_tx[TABLE_TYPE::USER_TABLE]->GetSnapshot();
+
+      assert(org_snapshot == m_read_opts[TABLE_TYPE::USER_TABLE].snapshot);
+      m_read_opts[TABLE_TYPE::USER_TABLE].snapshot = nullptr;
+
       rollback_to_stmt_savepoint();
 
       const rocksdb::Snapshot *const cur_snapshot =
```

laurynas-biveinis commented:
This patch is not enough. I'll remove the extra added accesses from other threads until the `m_read_opts[USER_TABLE].snapshot` concurrency is fixed

laurynas-biveinis commented:
@sunshine-Chun, addressed all comments as discussed above, fixed one linked list middle pointer manipulation logic error, and nanosecond to second conversion. Ready for review!

sunshine-Chun commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -5479,12 +5525,12 @@ class Rdb_transaction_impl : public Rdb_transaction {
   }
 
   virtual ~Rdb_transaction_impl() override {
+    rollback();
```
What will happen when we finish a transaction? My understanding is this transaction will be first reinsert without snapshot to the end of the list in on_commit(). Then when the transaction is deconstructed, it will be moved to the end again in the rollback(). Lastly, it will be removed from the list via erase(). Did I miss something, or we do have duplicated/redundant operations?

laurynas-biveinis replied:
The destructor is only executed on a session disconnect, where it may be called with or without going through commit/rollback first. While a session is alive, the transaction object does not go through destruction but is reused after completing rollback/commit.

luqun commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -3823,18 +3868,22 @@ class Rdb_transaction {
 
   void incr_update_count(TABLE_TYPE table_type) {
     assert(!is_ac_nl_ro_rc_transaction());
+    assert(table_type != TABLE_TYPE::USER_TABLE || get_snapshot_ts() != 0);
```
maybe add similar check into incr_insert_count()?

laurynas-biveinis replied:
That one might be called without snapshot during bulkloading. I
have added a comment there.

luqun commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -5794,55 +5840,179 @@ class Rdb_writebatch_impl : public Rdb_transaction {
+  do_erase(tx);
+  fixup_newest_snapshot_after_erase(tx);
+  tx->m_earliest_snapshot_ts = 0;
+  tx->next = nullptr;
```
why only clean tx->next? how about tx->prev?

laurynas-biveinis replied:
`do_insert_without_snapshot` handles `tx->prev`. It does not handle `tx->next` because the other caller sets it outside the mutex critical section, to have it as short as possible.

laurynas-biveinis commented:
>    void incr_update_count(TABLE_TYPE table_type) {
>      assert(!is_ac_nl_ro_rc_transaction());
> +    assert(table_type != TABLE_TYPE::USER_TABLE || 
> get_snapshot_ts() != 0);
>
> maybe add similar check into incr_insert_count()?

That one might be called without snapshot during bulkloading. I 
have added a comment there.

laurynas-biveinis commented:
Pushed the branch again. The only changes are an extra comment at `incr_insert_count` and an additional note in the commit message, no actual code changes, any reviews in progress should not be invalidated by this push.

sunshine-Chun commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -3925,15 +3976,15 @@ class Rdb_transaction {
+      if (new_earliest_ts == 0) new_earliest_ts = snapshot->GetUnixTime();
+      Rdb_transaction_list::reinsert_with_snapshot(this, new_earliest_ts);
+    } else {
+      assert(earliest_ts <= snapshot->GetUnixTime());
```
Abhinav left some comment here:
1. Can we use the mysql_bin_log.get_max_write_hlc_nsec() only when binlog ttl is enabled?
2. For the assertion, when binlog ttl is enabled, we are comparing hlc with wall clock, that doesn't seem correct.
Can we do something like
```
if (earliest_ts == 0) {
      auto new_earliest_ts =
          rdb_is_binlog_ttl_enabled() && mysql_bin_log.get_max_write_hlc_nsec()
              ? mysql_bin_log.get_max_write_hlc_nsec() / 1'000'000'000
              : snapshot->GetUnixTime();
      Rdb_transaction_list::reinsert_with_snapshot(this, new_earliest_ts);
    } else {
      assert(earliest_ts <= rdb_is_binlog_ttl_enabled() &&
                     mysql_bin_log.get_max_write_hlc_nsec()
                 ? mysql_bin_log.get_max_write_hlc_nsec() / 1'000'000'000
                 : snapshot->GetUnixTime());
    }
```

sunshine-Chun replied:
This is the last comment needs to be addressed before commit if Luqun and Herman don't have more comments.

laurynas-biveinis replied:
I am testing the following:
```c++
  [[nodiscard]] static std::int64_t binlog_hlc_or_snapshot_ts(
      const rocksdb::Snapshot &snapshot) {
    if (!rdb_is_binlog_ttl_enabled()) return snapshot.GetUnixTime();
    const auto binlog_max_write_hlc_ns = mysql_bin_log.get_max_write_hlc_nsec();
    return (binlog_max_write_hlc_ns != 0) ? binlog_max_write_hlc_ns
                                          : snapshot.GetUnixTime();
  }

  void snapshot_created(const rocksdb::Snapshot *const snapshot) {
    assert(snapshot != nullptr);

    m_read_opts[USER_TABLE].snapshot = snapshot;

    const auto earliest_ts = get_earliest_snapshot_ts();
    if (earliest_ts == 0) {
      const auto new_earliest_ts = binlog_hlc_or_snapshot_ts(*snapshot);
      Rdb_transaction_list::reinsert_with_snapshot(this, new_earliest_ts);
    } else {
      assert(earliest_ts <= binlog_hlc_or_snapshot_ts(*snapshot));
    }
    m_is_delayed_snapshot = false;
  }
```

laurynas-biveinis replied:
Ugh, of course with a division by a billion

laurynas-biveinis commented:
Addressed last review comments, ready for review again

sunshine-Chun commented:
rocksdb_binlog_ttl.test is failing for ```std::int64_t myrocks::Rdb_transaction::get_snapshot_ts() const: Assertion `m_earliest_snapshot_ts <= result || result == 0' failed.```

laurynas-biveinis commented:
> rocksdb_binlog_ttl.test is failing for `` std::int64_t myrocks::Rdb_transaction::get_snapshot_ts() const: Assertion `m_earliest_snapshot_ts <= result || result == 0' failed. ``

@sunshine-Chun , division by a billion got lost again :( Fixed
